### DISCUSSION TITLE
Delete web sqlite database from Dart

### DISCUSSION THREAD

**Discussion by davidmartos96** [simolus3/drift#2821] [Q&A] [answered]
Hello, I was wondering if it was possible to delete the database programatically on web. The same way you would do with the Chrome/Firefox devtools or with the Javascript API.

**Comment by simolus3** [comment]
Are you using `package:drift/wasm.dart`? That provides [`deleteDatabase`](https://drift.simonbinder.eu/api/drift.wasm/wasmproberesult/deletedatabase), available through `WasmDatabase.probe()`.

**Reply by davidmartos96** [reply]
Yes, I have this essentially, with a dart vm counter part.

```dart
import 'dart:async';

import 'package:drift/drift.dart';
import 'package:drift/wasm.dart';
import 'package:flutter/foundation.dart';

const webDbName = "todo-app";

/// Obtains a database connection for running drift on the web.
DatabaseConnection connect() {
  return DatabaseConnection.delayed(Future(() async {
    final db = await WasmDatabase.open(
      databaseName: webDbName,
      sqlite3Uri: Uri.parse('/sqlite3.wasm'),
      driftWorkerUri: Uri.parse('/drift_worker.js'),
    );

    if (db.missingFeatures.isNotEmpty) {
      debugPrint('Using ${db.chosenImplementation} due to unsupported '
          'browser features: ${db.missingFeatures}');
    }

    print("Chosen Web sqlite implementation: ${db.chosenImplementation}");

    return db.resolvedExecutor;
  }));
}

Future<void> deleteTodosDbFile() async {
  throw UnimplementedError();
}

```

**Reply by davidmartos96** [reply]
I've just tried the following, but it looks like the existingDatabases list is empty

```dart
Future<void> deleteTodosDbFile() async {
  final probeRes = await WasmDatabase.probe(
    sqlite3Uri: _sqlite3Uri,
    driftWorkerUri: _driftWorkerUri,
  );

  final db = _firstWhereOrNull(
    probeRes.existingDatabases,
    (dbInfo) => dbInfo.$2 == webDbName,
  );

  print(probeRes.existingDatabases);

  if (db != null) {
    await probeRes.deleteDatabase(db);
  }
}
```

**Reply by davidmartos96** [reply]
@simolus3 Found the problem, I wasn't providing the database name to probe. There is one additional issue, though.
The `probeRes.deleteDatabase` future never completes. From what I have researched, it is because for indexeddb, delete waits for all connections to close.
The problem is that even if drfitDb.close() is called, the shared worker still holds the native db connection. Is there any way to close it also in the worker?

**Reply by simolus3** [reply, marked as answer]
>  The problem is that even if drfitDb.close() is called, the shared worker still holds the native db connection. Is there any way to close it also in the worker?

I've implemented this in 09c6cf0b4e968204be5eda63614e815078fea970, but it also requires a newer worker. It's not 100% reliable (if a second tab opens a database connection and gets closed, it can't tell the database server that the connection is dead). But the common case of opening a database, closing it and then deleting it should work.

**Reply by davidmartos96** [reply]
@simolus3 Thanks Simon, I will try it out next week. If I understood correctly, the limitation is that the worker is not able to detect when a tab closes, so it cannot tell when the number of clients reaches 0?

**Reply by simolus3** [reply]
Exactly, and I'm not aware of a good workaround for that. Perhaps we could send pings periodically but I'm not sure whether that's worth it.

**Reply by miguelcmedeiros** [reply]
@simolus3 I was trying the approach mentioned above to delete a database on web, but I get the following error when `await probeRes.deleteDatabase(db)` is called:
```
InvalidStateError: Failed to read the 'error' property from 'IDBRequest': The request has not finished.
package:drift/src/web/wasm_setup/shared.dart 440:31                               <fn>
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 598:37  _checkAndCall
dart-sdk/lib/_internal/js_dev_runtime/private/profile.dart 117:39                 dcall
package:web/src/helpers/events/streams.dart 165:63                                <fn>
dart-sdk/lib/_internal/js_dev_runtime/patch/js_allow_interop_patch.dart 224:27    _callDartFunctionFast1
```

This happens when handling `blockedEvent` for `IDBRequest`, which in this case the pending event is `IDBVersionChangeEvent`.  This happens most of the times, but not always, so it seems to be a race-condition.

Before attempting to delete the database I'm closing it but it seems that according to [this doc](https://developer.mozilla.org/en-US/docs/Web/API/IDBOpenDBRequest/blocked_event), the connection might still be open.

What I also noticed is that event when the `blockedEvent` is triggered, the `successEvent` is also triggered afterwards. Also, from the stacktrace above, note that the `error` property is not set, which makes me wonder the request should complete with error when handling `blockedEvent`.

What do you think?

**Reply by simolus3** [reply]
> Also, from the stacktrace above, note that the `error` property is not set, which makes me wonder the request should complete with error when handling `blockedEvent`.

From a quick look at the spec, it's not clear to be if a blocked event is supposed to be terminal or if it can be followed by a success event. If we know that we'll get a blocked/error event after the blocked, we should probably just ignore the blocked event. But I don't think that's the case.

> What do you think?

It looks like deleting the database will trigger a version change event on other databases which have the option to close at that point. Maybe that's something worth doing.

**Reply by miguelcmedeiros** [reply]
> It looks like deleting the database will trigger a version change event on other databases which have the option to close at that point. Maybe that's something worth doing.

Not sure I get what you meant with this. Can you please expand on this?

**Reply by simolus3** [reply]
When we call [`deleteDatabase`](https://developer.mozilla.org/en-US/docs/Web/API/IDBFactory/deleteDatabase) while other connections to the database exists, those connections are supposed to receive an `versionchange` event. In the spec, it says that if these users all close their connections immediately after receiving that event, there will be no `blocked` event.

Given that there is a request to delete the database, I wonder if we should actively listen for the `versionchange` events after the database has already been opened just close the connection on that event (given that some other context is deleting the database). But it's probably a lot of work to pass these notifications all the way up to the user of the database.

**Reply by miguelcmedeiros** [reply]
> If we know that we'll get a blocked/error event after the blocked, we should probably just ignore the blocked event. But I don't think that's the case.

From the spec of [deleteDatabase](https://developer.mozilla.org/en-US/docs/Web/API/IDBFactory/deleteDatabase), it only mentions that a `success` or an `error` event is expected and since `error` parameter of `IDBRequest` is not set, I assume that we are safe with not completing with an error when a `blocked` event is triggered.

> Given that there is a request to delete the database, I wonder if we should actively listen for the versionchange events after the database has already been opened just close the connection on that event (given that some other context is deleting the database). But it's probably a lot of work to pass these notifications all the way up to the user of the database.

I noticed that when `IDBVersionChangeEvent` is emitted, it has some old version (e.g. 1) and `null` as the new version. I wonder if it's enough for the connections to listen to this event and auto-close when the new version is set to `null`.

**Reply by miguelcmedeiros** [reply]
Maybe it's good to share in which context I'm trying to delete the database on web. We noticed recently that many user that run our app on web ended up with a corrupted database, namely with error `SqliteException(11): while executing statement, database disk image is malformed`.

In order to help those users we want to deploy a solution where we run `PRAGMA quick_check` on a open database to detect corruption, close the database connection, delete the database and open a new database connection.
